### ISSUE TITLE
apply to my own datasets

### ISSUE THREAD

**Issue by summerwbb** [facebookresearch/multigrain#6]
hi! I'm glad to get your source code.！but how can I apply it to my own dataset?

**Comment by StepOITD** [comment]
@summerwbb hi there, did you figure out how to do that? Any help would be appreciated

**Comment by bermanmaxim** [comment]
Hi @summerwbb @StepOITD are you interested in using your own dataset at test-time or training time, and does your question concern retrieval or classification? In general you can extract features using the "normalized_embedding" output of the network (https://github.com/facebookresearch/multigrain/blob/eff92844f7bcd9fc82bdf7bf7f84ec08e6d590f5/multigrain/lib/multigrain.py#L83).

**Comment by StepOITD** [comment]
Hi @bermanmaxim , I'm looking for ways to train on my own dataset, and I think it would be really nice if you can give a little descrioption about how the dataset should be arranged in the README file, or I have to  go through the source code to reproduce the result step by step. Any way thank you for answering!

**Comment by bermanmaxim** [comment]
So if you follow the paper the network is trained on imagenet only and evaluated both on classification and retrieval datasets; moreover the imagenet training code is included. The retrieval aspect of our network is therefore unsupervised (trained with data-augmentation). Training on retrieval datasets together with the training of the classifier could be an interesting extension but we did not experiment with it.

**Comment by StepOITD** [comment]
Hi, @bermanmaxim, one more question, i notice that there's a NASNet-A-Mobile based model, joint_1B_1.0.pth, https://dl.fbaipublicfiles.com/multigrain/finetuned_models/nasnetamobile-finetune350.pth
So what does the "1.0" mean? I thought it was λ, which is the ratio of `classification loss` and `retr loss`, but that means the model was only trained using classification loss?
Or the model was fine-tuned only using classification loss? If so what was the λ while training?

**Comment by StepOITD** [comment]
Also, I wonder which feature is better for retrieval missions, cause one of the script is using `output_dict['embedding']` while another is using `output_dict['normalized_embedding']`

[evaluate.py](https://github.com/facebookresearch/multigrain/blob/master/scripts/evaluate.py#L90-#L94)

[whiten.py](https://github.com/facebookresearch/multigrain/blob/master/scripts/whiten.py#L73-#L79)

**Comment by bermanmaxim** [comment]
> Hi, @bermanmaxim, one more question, i notice that there's a NASNet-A-Mobile based model, joint_1B_1.0.pth, https://dl.fbaipublicfiles.com/multigrain/finetuned_models/nasnetamobile-finetune350.pth
> So what does the "1.0" mean? I thought it was λ, which is the ratio of `classification loss` and `retr loss`, but that means the model was only trained using classification loss?
> Or the model was fine-tuned only using classification loss? If so what was the λ while training?

Indeed, it was fine-tuned with classification loss only, because for these bigger networks we only fine-tuned the pooling exponent in the pooling layer of off-the-shelf classification networks for a given resolution - see Sec. 4 in supplementary. Training with retrieval loss should bring additional improvements both in terms of retrieval and classification, and should not require too many resources in the case of the mobile model (and all the training code in the repository should be compatible with that model).

**Comment by bermanmaxim** [comment]
> Also, I wonder which feature is better for retrieval missions, cause one of the script is using `output_dict['embedding']` while another is using `output_dict['normalized_embedding']`
> 
> [evaluate.py](https://github.com/facebookresearch/multigrain/blob/master/scripts/evaluate.py#L90-#L94)
> 
> [whiten.py](https://github.com/facebookresearch/multigrain/blob/master/scripts/whiten.py#L73-#L79)

In general it should be better to use the normalized embeddings, after having done the whitening step.

**Comment by StepOITD** [comment]
@bermanmaxim and that top-1 accuracies, in the table, are classification accuracies  or retrieval accuracies?

**Comment by saiaman** [comment]
Any infos on how to train on a custom retrieval dataset ??
